### PR TITLE
[quant] Add PT2 QAT tutorial

### PR DESCRIPTION
**Pull request by andrewor14**
Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Fixes #ISSUE_NUMBER

## Checklist

- [ ] The issue that is being fixed is referred in the description (see above "Fixes #ISSUE_NUMBER")
- [ ] Only one issue is addressed in this pull request
- [ ] Labels from the issue that this PR is fixing are added to this pull request
- [ ] No unnecessary issues are included into this pull request.

### CODE REVIEW

**Comment by andrewor14** [comment]
@jerryzh168 @svekars Please have a look. Thanks.

File: prototype_source/pt2e_quant_qat_static.rst [hunk 1]
```diff
@@ -0,0 +1,429 @@
...
+
+    quantized_model = convert_pt2e(prepared_model)
+
+    # move certain ops like dropout to eval mode, equivalent to `m.eval()`
+    torch.ao.quantization.move_exported_model_to_eval(m)
+
+    print(quantized_model)
+
+    top1, top5 = evaluate(quantized_model, criterion, data_loader_test, neval_batches=num_eval_batches)
+    print('Final evaluation accuracy on %d images, %2.2f' % (num_eval_batches * eval_batch_size, top1.avg))
```
Comment (jerryzh168) [inline prototype_source/pt2e_quant_qat_static.rst:418]: should we show the results here? and compare with baseline?

**Review by jerryzh168 — COMMENTED** [review]
should we add a section on how to save/load the QAT model as well

File: prototype_source/pt2e_quant_qat_static.rst (hunk 1 above)
Comment (andrewor14) [reply prototype_source/pt2e_quant_qat_static.rst:418]: I don't have the results right now. Can we land this first for the 2.1 blog post and add it later?

**Comment by andrewor14** [comment]
> should we add a section on how to save/load the QAT model as well

We can, but it would duplicate the serialization section in the PTQ tutorial. Should I add it anyway?

**Comment by jerryzh168** [comment]
> > should we add a section on how to save/load the QAT model as well
> 
> We can, but it would duplicate the serialization section in the PTQ tutorial. Should I add it anyway?

it should be different, I'm talking about how to save and load a model prepared for QAT here, not the converted model, so it should be saving the state_dict etc.

File: prototype_source/pt2e_quant_qat_static.rst (hunk 1 above)
Comment (jerryzh168) [reply prototype_source/pt2e_quant_qat_static.rst:418]: OK then maybe add a comment?

File: prototype_source/pt2e_quant_qat_static.rst [hunk 2]
```diff
@@ -0,0 +1,470 @@
...
+        # then you can just call `torch.ao.quantization.move_exported_model_to_eval/train`.
+        # However, the latter API is not ready yet and will be available in the near future.
+        if (nepoch + 1) % num_epochs_between_evals == 0:
+            prepared_model_copy = copy.deepcopy(prepared_model)
+            quantized_model = convert_pt2e(prepared_model_copy)
+            top1, top5 = evaluate(quantized_model, criterion, data_loader_test, neval_batches=num_eval_batches)
+            print('Epoch %d: Evaluation accuracy on %d images, %2.2f' % (nepoch, num_eval_batches * eval_batch_size, top1.avg))
+
+
+Saving and loading model checkpoints
```
Comment (jerryzh168) [inline prototype_source/pt2e_quant_qat_static.rst:399]:
```suggestion
Saving and Loading Model Checkpoints
```

**Review by jerryzh168 — APPROVED** [review]
LGTM thanks!

File: prototype_source/pt2e_quant_qat_static.rst [hunk 3]
```diff
@@ -0,0 +1,470 @@
...
+================================================================
+**Author**: `Andrew Or <https://github.com/andrewor14>`_
+
+This tutorial shows how to perform quantization-aware training (QAT) in
+graph mode based on `torch.export.export <https://pytorch.org/docs/main/export.html>`_.
+For more details about PyTorch 2.0 Export Quantization in general, refer
+to the `static post training quantization tutorial <https://pytorch.org/tutorials/prototype/pt2e_quant_ptq_static.html>`_.
+
+The PyTorch 2.0 Export QAT flow looks like the following. It is similar
+to the post training quantization (PTQ) flow for the most part:
```
Comment (svekars) [inline prototype_source/pt2e_quant_qat_static.rst:11]:
```suggestion
The PyTorch 2.0 Export QAT flow looks like the following—it is similar
to the post training quantization (PTQ) flow for the most part:
```

File: prototype_source/pt2e_quant_qat_static.rst [hunk 4]
```diff
@@ -0,0 +1,470 @@
...
+  # we have a model with aten ops doing integer computations when possible
+
+  # move the quantized model to eval mode, equivalent to `m.eval()`
+  torch.ao.quantization.move_exported_model_to_eval(m)
+
+Note that calling `model.eval()` or `model.train()` after program capture is
+not allowed, because these methods no longer correctly change the behavior of
+certain ops like dropout and batch normalization. Instead, please use
+`torch.ao.quantization.move_exported_model_to_eval()` and
+`torch.ao.quantization.move_exported_model_to_train()` (coming soon)
```
Comment (svekars) [inline prototype_source/pt2e_quant_qat_static.rst:62]:
```suggestion
``torch.ao.quantization.move_exported_model_to_eval()`` and
``torch.ao.quantization.move_exported_model_to_train()`` (coming soon)
```

File: prototype_source/pt2e_quant_qat_static.rst [hunk 5]
```diff
@@ -0,0 +1,470 @@
...
+
+  # train omitted
+
+  m = convert_pt2e(m)
+  # we have a model with aten ops doing integer computations when possible
+
+  # move the quantized model to eval mode, equivalent to `m.eval()`
+  torch.ao.quantization.move_exported_model_to_eval(m)
+
+Note that calling `model.eval()` or `model.train()` after program capture is
```
Comment (svekars) [inline prototype_source/pt2e_quant_qat_static.rst:58]:
```suggestion
Note that calling ``model.eval()`` or ``model.train()`` after program capture is
```

File: prototype_source/pt2e_quant_qat_static.rst [hunk 6]
```diff
@@ -0,0 +1,470 @@
...
+not allowed, because these methods no longer correctly change the behavior of
+certain ops like dropout and batch normalization. Instead, please use
+`torch.ao.quantization.move_exported_model_to_eval()` and
+`torch.ao.quantization.move_exported_model_to_train()` (coming soon)
+respectively.
+
+
+Define Helper Functions and Prepare Dataset
```
Comment (svekars) [inline prototype_source/pt2e_quant_qat_static.rst:66]:
```suggestion
Define Helper Functions and Prepare the Dataset
```

File: prototype_source/pt2e_quant_qat_static.rst [hunk 7]
```diff
@@ -0,0 +1,470 @@
...
+-------------------------------------------
+
+We’ll start by doing the necessary imports, defining some helper functions and
+prepare the data. These steps are very similar to the ones defined in the
+`static post training quantization tutorial <https://pytorch.org/tutorials/advanced/static_quantization_tutorial.html>`_.
+
+To run the code in this tutorial using the entire ImageNet dataset, first
+download Imagenet by following the instructions at here
```
Comment (svekars) [inline prototype_source/pt2e_quant_qat_static.rst:74]:
```suggestion
download ImageNet by following the instructions in
```

File: prototype_source/pt2e_quant_qat_static.rst [hunk 8]
```diff
@@ -0,0 +1,470 @@
...
+`ImageNet Data <http://www.image-net.org/download>`_. Unzip the downloaded file
+into the ``data_path`` folder.
+
+Download the `torchvision resnet18 model <https://download.pytorch.org/models/resnet18-f37072fd.pth>`_
```
Comment (svekars) [inline prototype_source/pt2e_quant_qat_static.rst:78]:
```suggestion
Next, ownload the `torchvision resnet18 model <https://download.pytorch.org/models/resnet18-f37072fd.pth>`_
```
You had "first" above, you need to have a "next" =)

File: prototype_source/pt2e_quant_qat_static.rst [hunk 9]
```diff
@@ -0,0 +1,470 @@
...
+and rename it to ``data/resnet18_pretrained_float.pth``.
+
```
Comment (svekars) [inline prototype_source/pt2e_quant_qat_static.rst:80]: You might want to add something like: In the following code block, we do blah... and describe what is happening there.

File: prototype_source/pt2e_quant_qat_static.rst [hunk 10]
```diff
@@ -0,0 +1,470 @@
...
+    from torch._export import dynamic_dim
+
+    example_inputs = (torch.rand(2, 3, 224, 224),)
+    exported_model = capture_pre_autograd_graph(
+        float_model,
+        example_inputs,
+        constraints=[dynamic_dim(example_inputs[0], 0)],
+    )
+
+``capture_pre_autograd_graph`` is a short term API, it will be updated to use the offical ``torch.export`` API when that is ready.
```
Comment (svekars) [inline prototype_source/pt2e_quant_qat_static.rst:290]:
```suggestion
.. note::

   ``capture_pre_autograd_graph`` is a short term API, it will be updated to use the offical ``torch.export`` API when that is ready.
```

File: prototype_source/pt2e_quant_qat_static.rst [hunk 11]
```diff
@@ -0,0 +1,470 @@
...
+
+
+Import the Backend Specific Quantizer and Configure how to Quantize the Model
+-----------------------------------------------------------------------------
+
+The following code snippets describes how to quantize the model:
```
Comment (svekars) [inline prototype_source/pt2e_quant_qat_static.rst:296]:
```suggestion
The following code snippets describe how to quantize the model:
```

File: prototype_source/pt2e_quant_qat_static.rst [hunk 12]
```diff
@@ -0,0 +1,470 @@
...
+.. code-block:: python
+
+    prepared_model = prepare_qat_pt2e(exported_model, quantizer)
+    print(prepared_model)
+
+.. note::
+
+    If your model contains batch normalization, the actual ATen ops you get
+    in the graph depend on the model's device when you export the model.
+    If the model is on CPU, then you'll get `torch.ops.aten._native_batch_norm_legit`.
```
Comment (svekars) [inline prototype_source/pt2e_quant_qat_static.rst:334]:
```suggestion
    If the model is on CPU, then you'll get ``torch.ops.aten._native_batch_norm_legit``.
```

File: prototype_source/pt2e_quant_qat_static.rst [hunk 13]
```diff
@@ -0,0 +1,470 @@
...
+    If the model is on CUDA, then you'll get `torch.ops.aten.cudnn_batch_norm`.
```
Comment (svekars) [inline prototype_source/pt2e_quant_qat_static.rst:335]:
```suggestion
    If the model is on CUDA, then you'll get ``torch.ops.aten.cudnn_batch_norm``.
```

File: prototype_source/pt2e_quant_qat_static.rst [hunk 14]
```diff
@@ -0,0 +1,470 @@
...
+    However, this is not fundamental and may be subject to change in the future.
+
+    Between these two ops, it has been shown that `torch.ops.aten.cudnn_batch_norm`
```
Comment (svekars) [inline prototype_source/pt2e_quant_qat_static.rst:338]:
```suggestion
    Between these two ops, it has been shown that ``torch.ops.aten.cudnn_batch_norm``
```

File: prototype_source/pt2e_quant_qat_static.rst [hunk 15]
```diff
@@ -0,0 +1,470 @@
...
+    provides better numerics on models like MobileNetV2. To get this op, either
+    call `model.cuda()` before export, or run the following after prepare to manually
```
Comment (svekars) [inline prototype_source/pt2e_quant_qat_static.rst:340]:
```suggestion
    call ``model.cuda()`` before export, or run the following after prepare to manually
```

File: prototype_source/pt2e_quant_qat_static.rst [hunk 16]
```diff
@@ -0,0 +1,470 @@
...
+            if n.target == torch.ops.aten._native_batch_norm_legit.default:
+                n.target = torch.ops.aten.cudnn_batch_norm.default
+        prepared_model.recompile()
+
+    In the future, we plan to consolidate the batch normalization ops such that
+    the above will no longer be necessary.
+
+Training Loop
+-----------------------------------------------------------------------------
+
```
Comment (svekars) [inline prototype_source/pt2e_quant_qat_static.rst:355]: Need some intro sentence here that explains the code below

File: prototype_source/pt2e_quant_qat_static.rst [hunk 17]
```diff
@@ -0,0 +1,470 @@
...
+    # resume training or perform inference
+
+
+Convert the Trained Model to a Quantized Model
+----------------------------------------------------------
+
+``convert_pt2e`` takes a calibrated model and produces a quantized model.
+Note that, before inference, you must first call
+`torch.ao.quantization.move_exported_model_to_eval()` to ensure certain ops
```
Comment (svekars) [inline prototype_source/pt2e_quant_qat_static.rst:443]:
```suggestion
``torch.ao.quantization.move_exported_model_to_eval()`` to ensure certain ops
```

**Review by svekars — COMMENTED** [review]
Looks good, a few minor editorial nits.

**Comment by andrewor14** [comment]
Ok, I think I addressed everyone's comments. @svekars is this good to merge?

**Comment by svekars** [comment]
@andrewor14 we will merge next week.

**Review by svekars — APPROVED** [review]

File: prototype_source/prototype_index.rst [hunk 18]
```diff
@@ -82,6 +82,13 @@ Prototype features are not available as part of binary distributions like PyPI o
    :link: ../prototype/pt2e_quant_ptq_static.html
    :tags: Quantization          
 
+.. customcarditem::
+   :header: PyTorch 2.0 Export Static Quantization-Aware Training
```
Comment (jerryzh168) [inline prototype_source/prototype_index.rst:86]: nit: maybe we can remove `Static`, and also change 2.0 to 2

File: prototype_source/prototype_index.rst [hunk 19]
```diff
@@ -82,6 +82,13 @@ Prototype features are not available as part of binary distributions like PyPI o
...
+   :card_description: Learn how to use static Quantization-Aware-Training in PyTorch 2.0 Export.
+   :image: ../_static/img/thumbnails/cropped/generic-pytorch-logo.png
+   :link: ../prototype/pt2e_quant_qat_static.html
```
Comment (jerryzh168) [inline prototype_source/prototype_index.rst:89]: I think we can remove `_static` here as well

File: prototype_source/prototype_index.rst [hunk 20]
```diff
@@ -82,6 +82,13 @@ Prototype features are not available as part of binary distributions like PyPI o
    :link: ../prototype/pt2e_quant_ptq_static.html
    :tags: Quantization          
 
+.. customcarditem::
+   :header: PyTorch 2.0 Export Static Quantization-Aware Training
+   :card_description: Learn how to use static Quantization-Aware-Training in PyTorch 2.0 Export.
```
Comment (jerryzh168) [inline prototype_source/prototype_index.rst:87]: and here

File: prototype_source/pt2e_quant_qat.rst [hunk 21]
```diff
@@ -0,0 +1,476 @@
+(prototype) PyTorch 2.0 Export Quantization-Aware Training (QAT)
```
Comment (jerryzh168) [inline prototype_source/pt2e_quant_qat.rst:1]: please don't forget to update 2.0 to 2 in this file

File: prototype_source/pt2e_quant_qat.rst [hunk 22]
```diff
@@ -0,0 +1,476 @@
...
+
+Conclusion
+--------------
+
+In this tutorial, we demonstrated how to run Quantization-Aware Training (QAT)
+flow in PyTorch 2.0 Export Quantization. After convert, the rest of the flow
+is the same as Post-Training Quantization (PTQ); the user can
+serialize/deserialize the model and further lower it to a backend that supports
+inference with XNNPACK backend. For more detail, follow the
+`static PTQ tutorial <https://pytorch.org/tutorials/prototype/pt2e_quant_ptq_static.html>`_.
```
Comment (jerryzh168) [inline prototype_source/pt2e_quant_qat.rst:476]: and remove static

File: prototype_source/pt2e_quant_qat.rst (hunk 21 above)
Comment (andrewor14) [reply prototype_source/pt2e_quant_qat.rst]: oh whoops

File: prototype_source/pt2e_quant_qat.rst [hunk 23]
```diff
@@ -0,0 +1,476 @@
+(prototype) PyTorch 2 Export Quantization-Aware Training (QAT)
+================================================================
+**Author**: `Andrew Or <https://github.com/andrewor14>`_
+
+This tutorial shows how to perform quantization-aware training (QAT) in
+graph mode based on `torch.export.export <https://pytorch.org/docs/main/export.html>`_.
+For more details about PyTorch 2 Export Quantization in general, refer
+to the `post training quantization tutorial <https://pytorch.org/tutorials/prototype/pt2e_quant_ptq.html>`_.
```
Comment (svekars) [inline prototype_source/pt2e_quant_qat.rst:8]: this link is 404. Should it be:
```suggestion
to the `post training quantization tutorial <https://pytorch.org/tutorials/prototype/pt2e_quant_ptq_static.html>`_.
```

File: prototype_source/pt2e_quant_qat.rst [hunk 24]
```diff
@@ -0,0 +1,476 @@
...
+Note that calling ``model.eval()`` or ``model.train()`` after program capture is
+not allowed, because these methods no longer correctly change the behavior of
+certain ops like dropout and batch normalization. Instead, please use
+``torch.ao.quantization.move_exported_model_to_eval()`` and
+``torch.ao.quantization.move_exported_model_to_train()`` (coming soon)
+respectively.
+
+
+Define Helper Functions and Prepare the Dataset
+-------------------------------------------
```
Comment (svekars) [inline prototype_source/pt2e_quant_qat.rst:67]:
```suggestion
----------------------------------------------------
```

File: prototype_source/pt2e_quant_qat.rst [hunk 25]
```diff
@@ -0,0 +1,476 @@
...
+
+Conclusion
+--------------
+
+In this tutorial, we demonstrated how to run Quantization-Aware Training (QAT)
+flow in PyTorch 2 Export Quantization. After convert, the rest of the flow
+is the same as Post-Training Quantization (PTQ); the user can
+serialize/deserialize the model and further lower it to a backend that supports
+inference with XNNPACK backend. For more detail, follow the
+`PTQ tutorial <https://pytorch.org/tutorials/prototype/pt2e_quant_ptq.html>`_.
```
Comment (svekars) [inline prototype_source/pt2e_quant_qat.rst:476]: this link is 404. Should it be:
```suggestion
`PTQ tutorial <https://pytorch.org/tutorials/prototype/pt2e_quant_ptq_static.html>`_.
```

**Review by svekars — APPROVED** [review]
